contracts/cw20-streams/src/contract.rs

```diff
@@ -0,0 +1,446 @@
+        rate_per_second: stream.rate_per_second,
+        start_time: stream.start_time,
+        end_time: stream.end_time,
+    })
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use cosmwasm_std::testing::{mock_dependencies, mock_env, mock_info, MOCK_CONTRACT_ADDR};
+    use cosmwasm_std::{Addr, CosmosMsg, WasmMsg};
+
+    #[test]
+    fn initialization() {
+        let mut deps = mock_dependencies();
+        let msg = InstantiateMsg {
+            owner: None,
+            cw20_addr: String::from(MOCK_CONTRACT_ADDR),
+        };
+
+        let info = mock_info("creator", &[]);
+        instantiate(deps.as_mut(), mock_env(), info, msg).unwrap();
+
+        let msg = QueryMsg::GetConfig {};
+        let res = query(deps.as_ref(), mock_env(), msg).unwrap();
+        let config: Config = from_binary(&res).unwrap();
+
+        assert_eq!(
+            config,
+            Config {
+                owner: Addr::unchecked("creator"),
+                cw20_addr: Addr::unchecked(MOCK_CONTRACT_ADDR)
+            }
+        );
+    }
+
+    #[test]
+    fn try_withdraw() {
+        let mut deps = mock_dependencies();
+        let msg = InstantiateMsg {
+            owner: None,
+            cw20_addr: String::from(MOCK_CONTRACT_ADDR),
+        };
+        let mut info = mock_info(MOCK_CONTRACT_ADDR, &[]);
+        instantiate(deps.as_mut(), mock_env(), info.clone(), msg).unwrap();
+
+        let sender = Addr::unchecked("Alice").to_string();
+        let recipient = Addr::unchecked("Bob").to_string();
+        let amount = Uint128::new(200);
+        let mut env = mock_env();
+        let start_time = env.block.time.plus_seconds(100).seconds();
+        let end_time = env.block.time.plus_seconds(300).seconds();
+
+        let msg = ExecuteMsg::Receive(Cw20ReceiveMsg {
+            sender,
+            amount,
+            msg: to_binary(&ReceiveMsg::CreateStream {
+                recipient,
+                start_time,
+                end_time,
+            })
+            .unwrap(),
+        });
+        execute(deps.as_mut(), env.clone(), info.clone(), msg).unwrap();
+
+        let msg = QueryMsg::GetStream { id: 1 };
+        let res = query(deps.as_ref(), mock_env(), msg).unwrap();
+        let stream: Stream = from_binary(&res).unwrap();
+
+        assert_eq!(
+            stream,
+            Stream {
+                owner: Addr::unchecked("Alice"),
+                recipient: Addr::unchecked("Bob"),
+                amount,
+                claimed_amount: Uint128::new(0),
+                start_time,
+                rate_per_second: Uint128::new(1),
+                end_time
+            }
+        );
+
+        let msg = ExecuteMsg::Withdraw { id: 1 };
+        info.sender = Addr::unchecked("Bob");
+        env.block.time = env.block.time.plus_seconds(150);
+        let res = execute(deps.as_mut(), env, info, msg).unwrap();
+        let msg = res.messages[0].clone().msg;
+
+        assert_eq!(
```

Comment:
This is a good basic check.

After this, I would add a second withdraw where the block_time > end_time.

Ensure that all remaining tokens are released and no more. 

`env.block.time = env.block.time.plus_seconds(500);`

expected:

```rust
     msg: to_binary(&Cw20ExecuteMsg::Transfer {
                     recipient: String::from("Bob"),
                     /// 200 total, 50 released earlier, pay out all that was left
                     amount: Uint128::new(150)
                 })
```

Reply:
This test will fail with the current code, I made suggestions how to fix the withdraw logic above.

Reply:
Done